### PR TITLE
Smart Spec Macro

### PR DESCRIPTION
**Pull request by kitlangton**
- [x] Implement for Scala 2
- [ ] Implement for Scala 3

A macro `suiteAll` which allows a MutableSpec like syntax, but simply rewrites the following:
```scala
  def spec =
    suiteAll("SweetSpec!") {
      val hello = "hello"

      test("test")(
        assertTrue(hello.startsWith("h"))
      )

      val cool = 123

      suiteAll("Nested") {
        test("test 2")(
          ZIO.service[Int].map { int =>
            assertTrue(cool == int)
          }
        )

        test("test 3")(
          ZIO.service[Int].map { int =>
            assertTrue(cool == int)
          }
        )
      }
    }
      .provide(ZLayer.succeed(123))
```

into

```scala
  def spec =
    suite("SweetSpec!") {
      val hello = "hello"

      val $test1 = test("test")(
        assertTrue(hello.startsWith("h"))
      )

      val cool = 123

      val $suite1 = suite("Nested") {
        val $test2 = test("test 2")(
          ZIO.service[Int].map { int =>
            assertTrue(cool == int)
          }
        )

        val $test3 = test("test 3")(
          ZIO.service[Int].map { int =>
            assertTrue(cool == int)
          }
        )

        test2 + $test3
      }

      $test1 + $suite1
    }
      .provide(ZLayer.succeed(123))
```

This has to be a whitebox macro, so IntelliJ will not know the full type. However, when used to define a single suite inside of `def spec`, the compiler error fired by #6669 will still fire (as well as when calling `.provide` without all the required layers).

### CODE REVIEW

File: test/shared/src/main/scala/zio/test/ZIOSpec.scala [hunk 1]
```diff
@@ -35,9 +35,12 @@ abstract class ZIOSpec[R: EnvironmentTag] extends ZIOSpecAbstract { self =>
   ): testConstructor.Out =
     zio.test.test(label)(assertion)
 
-  def suite[In](label: String)(specs: In*)(implicit
+  def suite[In](label: String)(spec: In, specs: In*)(implicit
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/ZIOSpec.scala:38]: I think we can revert this change now.

File: test-tests/shared/src/test/scala/zio/test/SweetSpec.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package zio.test
+import zio._
+
+object SweetSpec extends ZIOSpecDefault {
```
Comment (adamgfraser) [inline test-tests/shared/src/test/scala/zio/test/SweetSpec.scala:4]: Update name?

**Review by adamgfraser — APPROVED** [review]
Minor comments but otherwise good from my end when ready!

**Review by adamgfraser — APPROVED** [review]

**Comment by hmemcpy** [comment]
As heavy users of `MutableRunnableSpec` we thank you from the bottom of our type system :D
